### PR TITLE
Restore functionality

### PR DESCRIPTION
**Pull request by PizieDust**
This PR restores most of the functionality from the Mirage retreat. 
- [x] Create a unikernel
- [x] Destroy a unikernel
- [x] View a specific unikernel
- [ ] Console output

### CODE REVIEW

File: unikernel.ml [hunk 1]
```diff
@@ -147,14 +142,71 @@ struct
...
+                (let now = Ptime.v (P.now_d_ps ()) in
+                 let _, (t : Storage.t) = !store in
+                 let users = User_model.create_user_session_map t.users in
+                 let middlewares =
+                   [
+                     (* Middleware.email_verified_middleware now u:sers;*)
+                     Middleware.auth_middleware now users;
+                   ]
+                 in
+                 (* TODO: middleware, extract domain from middleware *)
```
Comment (hannesm) [inline unikernel.ml:157]:
```suggestion
```

this TODO is done below :)

File: unikernel.ml [hunk 2]
```diff
@@ -764,26 +816,85 @@ struct
...
+                let users = User_model.create_user_session_map t.users in
+                let middlewares = [ Middleware.auth_middleware now users ] in
+                Middleware.apply_middleware middlewares
+                  (fun _reqd ->
+                    match Middleware.user_of_cookie users now reqd with
+                    | Ok user -> (
+                        let unikernel_name =
+                          String.sub path 20 (String.length path - 20)
+                        in
+                        (* TODO: middleware, extract domain from middleware *)
```
Comment (hannesm) [inline unikernel.ml:851]:
```suggestion
```

also done now below.

File: unikernel.ml [hunk 3]
```diff
@@ -800,75 +911,173 @@ struct
...
+                    let ct =
+                      Multipart_form.Content_type.of_string
+                        (content_type ^ "\r\n")
+                    in
+                    match ct with
+                    | Error (`Msg msg) ->
+                        Logs.warn (fun m -> m "couldn't content-type: %s" msg);
+                        let status =
+                          {
+                            Utils.Status.code = 403;
```
Comment (hannesm) [inline unikernel.ml:954]: is 403 the right one here? isn't it rather 400 - bad request?

File: unikernel.ml [hunk 4]
```diff
@@ -800,75 +911,173 @@ struct
...
+                          (reply ~content_type:"application/json"
+                             (Utils.Status.to_json status))
+                    | Ok ct -> (
+                        match Multipart_form.of_string_to_list data ct with
                         | Error (`Msg msg) ->
-                            Logs.warn (fun m -> m "couldn't decode data %s" msg);
+                            Logs.warn (fun m -> m "couldn't multipart: %s" msg);
+                            let status =
+                              {
+                                Utils.Status.code = 403;
```
Comment (hannesm) [inline unikernel.ml:969]: same, I'd expect 400.

File: unikernel.ml [hunk 5]
```diff
@@ -800,75 +911,173 @@ struct
...
+                                Logs.info (fun m -> m "args %s" args);
+                                match Albatross_json.config_of_json args with
+                                | Ok cfg -> (
+                                    let config =
+                                      {
+                                        cfg with
+                                        image = Cstruct.of_string binary;
+                                      }
+                                    in
+                                    (* TODO: middleware, extract domain from middleware *)
```
Comment (hannesm) [inline unikernel.ml:996]: would be great to get this TODO fixed :) should be the same code as above.

File: unikernel.ml [hunk 6]
```diff
@@ -800,75 +911,173 @@ struct
...
+                                          }
+                                        in
+                                        reply ~content_type:"application/json"
+                                          (Utils.Status.to_json status))
+                                | Error (`Msg msg) ->
+                                    Logs.warn (fun m ->
+                                        m "couldn't decode data %s" msg);
+                                    let status =
+                                      {
+                                        Utils.Status.code = 403;
```
Comment (hannesm) [inline unikernel.ml:1046]: 400

File: unikernel.ml [hunk 7]
```diff
@@ -800,75 +911,173 @@ struct
...
+                                      }
+                                    in
+                                    Lwt.return
+                                      (reply ~content_type:"application/json"
+                                         (Utils.Status.to_json status)))
+                            | _ ->
+                                Logs.warn (fun m -> m "couldn't find fields");
+                                let status =
+                                  {
+                                    Utils.Status.code = 403;
```
Comment (hannesm) [inline unikernel.ml:1059]: 400

File: unikernel.ml (hunk 3 above)
Comment (reynir) [reply unikernel.ml:954]: I tend to agree - the client sent malformed request body

**Comment by hannesm** [comment]
Thanks a lot, I'm merging and will likely soon have some refactoring PR :)

**Pull request merged**